runtime/domain.c

```diff
@@ -1106,6 +1114,55 @@ int caml_domain_is_in_stw(void) {
 }
 #endif
 
+/* During a stop-the-world (STW), all currently running domains stop
+   their usual work and synchronize to all call the same function.
+
+   STW sections use [all_domains_lock] and the variable [stw_leader]
+   (0 when no STW section is running, the dom_internal* pointer of the
+   STW leader when a STW section is running) to guarantee that no
+   domain is running something else:
+
+   - If two STW sections are attempted in parallel, only one will
+     manage to take the lock, and the domain starting the other will
+     join that winning STW section, without running its own STW code
+     at all. (This is the [_try_] in the function name: if it returns
+     0, the STW section did not run at all, so you should call this
+     function in a loop.)
+
+   - Domain initialization code from [create_domain] will not run in
+     parallel with a STW section, as [create_domain] starts bit
+     looping until (1) it has the [all_domains_lock] and (2) there is
+     no current STW section (using the [stw_leader] variable).
+
+   - Domain cleanup code will not run in parallel with a STW section:
```

Comment:
I don't think that is strictly true. Domain cleanup code will run after the domain is no longer in the STW participant set. 
```suggestion
   - Domain cleanup code runs after the terminating domain has safely removed
     itself from the STW participant set:
```